### PR TITLE
Add a limit to the number of columns in the CLUSTERED BY clause

### PR DESCRIPTION
**Pull request by LakshSingla**
If there is a huge number of columns passed to the clustered by clause while ingesting via MSQ, then the Worker tasks can OOM. (With sequential merge in place, controller tasks shouldn't OOM).
This PR adds a limit to the number of clustered by columns that can be passed in a query and throws a fault in case they are exceeded. 

#### Release note
There is a limit to the number of columns that can be passed in the CLUSTERED BY clause while ingesting via MSQ.

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/QueryValidator.java
Comment (adarshsanjeev) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/QueryValidator.java:62]: Does only the final stage lead to an OOM? Wouldn't it be possible for more cluster by columns to be present in earlier stages than the final one?

**Review by adarshsanjeev — COMMENTED** [review]
Thanks for the PR! Had one comment.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java:29]: Let's document this fault as well.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/QueryValidator.java
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/QueryValidator.java:62]: The cluster by columns in the earlier stages might not have a 1:1 correspondence with the query that the user has written therefore raising a cluster by error, in that case, shouldn't be actionable for the user IMO. Hence I only added the limit in the final stage (the original query that the user has written). Along with the Sequential merge mode on, I think that there should be enough guard rails in place to prevent an OOM.

However we can add a limit on the cluster by in the other stages if we rephrase the error message as something like "Enough grouping keys present in stage [xx], the query might OOM". Those cluster by keys can correspond to something present in the group by clause for example. WDYT?

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/QueryValidator.java:62]: Looking at the TooManyColumnsFault, I think that we can also go ahead with the second proposition since that is also imposed at a per-stage level, which might not correspond to the final result that the user expects. (The wording might need to change though).

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java:29]: Thanks for pointing it out, updated!

**Review by adarshsanjeev — APPROVED** [review]
LGTM after resolving merge conflict!

**Review by cryptoe — APPROVED** [review]
LGTM. Will merge post the conflicts are resolved. 
Thanks @LakshSingla

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java:28]: We might need to add this to MSQIndexingModule.java

**Comment by LakshSingla** [comment]
Test failures seem unrelated/flaky, can the second stage of the CI/CD be run again?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TooManyClusteredByColumnsFault.java:28]: Thanks for pointing it out, I added it to the module.

**Comment by cryptoe** [comment]
Failure look unrelated. 
Thanks for the PR @LakshSingla.

**Pull request merged**
